core/src/Streamly/Internal/Data/Serialize/TH.hs

```diff
@@ -136,8 +144,185 @@ isUnitType _ = False
+        x | x < 8 -> [|$(go32 0) == zeroBits|]
+        _ -> [|$(go64 0) == zeroBits|]
+  where
+    tagLen = length tag
+    go8 i | i >= tagLen = [|zeroBits|]
+    go8 i = do
+        let wIntegral = litIntegral i
+        [|xor (unsafeInlineIO
+                   (Unbox.peekByteIndex
+                        ($(varE off) + $(litIntegral i))
+                        $(varE arr)))
+              ($(wIntegral) :: Word8) .|.
+          $(go8 (i + 1))|]
+    go16 i
+        | i >= tagLen = [|zeroBits|]
+    go16 i
+        | tagLen - i < 2 = go16 (tagLen - 2)
+    go16 i = do
+        let wIntegral =
+                litIntegral
+                    (shiftAdd w8_w16 [tag !! i, tag !! (i + 1)] :: Word16)
+        [|xor (unsafeInlineIO
+                   (Unbox.peekByteIndex
+                        ($(varE off) + $(litIntegral i))
+                        $(varE arr)))
+              ($(wIntegral) :: Word16) .|.
+          $(go16 (i + 2))|]
+    go32 i
+        | i >= tagLen = [|zeroBits|]
+    go32 i
+        | tagLen - i < 4 = go32 (tagLen - 4)
+    go32 i = do
+        let wIntegral =
+                litIntegral
+                    (shiftAdd
+                         w8_w32
+                         [ tag !! i
+                         , tag !! (i + 1)
+                         , tag !! (i + 2)
+                         , tag !! (i + 3)
+                         ] :: Word32)
+        [|xor (unsafeInlineIO
+                   (Unbox.peekByteIndex
+                        ($(varE off) + $(litIntegral i))
+                        $(varE arr)))
+              ($(wIntegral) :: Word32) .|.
+          $(go32 (i + 4))|]
+    go64 i
+        | i >= tagLen = [|zeroBits|]
+    go64 i
+        | tagLen - i < 8 = go64 (tagLen - 8)
+    go64 i = do
+        let wIntegral =
+                litIntegral
+                    (shiftAdd
+                         w8_w64
+                         [ tag !! i
+                         , tag !! (i + 1)
+                         , tag !! (i + 2)
+                         , tag !! (i + 3)
+                         , tag !! (i + 4)
+                         , tag !! (i + 5)
+                         , tag !! (i + 6)
+                         , tag !! (i + 7)
+                         ])
+        [|xor (unsafeInlineIO
+                   (Unbox.peekByteIndex
+                        ($(varE off) + $(litIntegral i))
+                        $(varE arr)))
+              ($(wIntegral) :: Word64) .|.
+          $(go64 (i + 8))|]
+
+--------------------------------------------------------------------------------
+-- Primitive serialization
+--------------------------------------------------------------------------------
+
+-- XXX Batch serialize? See xorCmp
+serializeW8List :: Name -> Name -> [Word8] -> Q Exp
+serializeW8List off arr w8List = do
+    [|let $(varP (makeN 0)) = $(varE off)
+       in $(doE (fmap makeBind [0 .. (lenW8List - 1)] ++
+                 [noBindS ([|pure $(varE (makeN lenW8List))|])]))|]
```

Comment:
May be too much code bloat. Not sure if this could be a real issue. Can we use a loop with the loop body unrolled?

Reply:
I'm going to add this as a developer comment. We can fix this if any issue arises.